--- a/drivers/gpu/drm/i915/i915_drv.c
+++ b/drivers/gpu/drm/i915/i915_drv.c
@@ -290,6 +290,9 @@
 	{0, 0, 0}
 };
 
+static void intel_detect_pch_(struct drm_i915_private *dev_priv,
+						 struct pci_dev *pch);
+
 #if defined(CONFIG_DRM_I915_KMS)
 MODULE_DEVICE_TABLE(pci, pciidlist);
 #endif
@@ -311,6 +314,26 @@
 	 * underneath. This is a requirement from virtualization team.
 	 */
 	pch = pci_get_class(PCI_CLASS_BRIDGE_ISA << 8, NULL);
+
+	dev_priv->pch_type = ~0;
+
+	while (pch)
+	{
+		intel_detect_pch_(dev_priv, pch);
+		if (dev_priv->pch_type != ~0)
+			break;
+		pch = pci_get_class(PCI_CLASS_BRIDGE_ISA << 8, pch);
+	}
+
+	if (pch)
+		pci_dev_put(pch);
+	else
+		dev_priv->pch_type = 0;
+}
+
+static void intel_detect_pch_(struct drm_i915_private *dev_priv,
+							  struct pci_dev *pch)
+{
 	if (pch) {
 		if (pch->vendor == PCI_VENDOR_ID_INTEL) {
 			int id;
@@ -328,7 +351,6 @@
 				DRM_DEBUG_KMS("Found PatherPoint PCH\n");
 			}
 		}
-		pci_dev_put(pch);
 	}
 }
 
